src/cms/logic.py

```diff
@@ -0,0 +1,323 @@
+from django.core.files.base import ContentFile
+
+from cms import models as models
+from journal import models as journal_models
+from press import models as press_models
+from repository import models as repository_models
+from utils.function_cache import cache
+from utils.logger import get_logger
+
+logger = get_logger(__name__)
+
+SITE_SEARCH_PATH = os.path.join(
+    settings.MEDIA_ROOT,
+    'press',
+    settings.SITE_SEARCH_DIR,
+)
+
+search_documents = []
+fetched_urls = set()
+
+
+def add_search_index_document(url, name, text):
+    """
+    Adds the data for a part or page to the search index
+    """
+    data = {}
+    data['id'] = len(search_documents)
+    data['url'] = url
+    data['name'] = name
+    data['text'] = text
+    search_documents.append(data)
+
+
+def gobble_sibling_text(sibling, original_part):
+    """
+    Recursively collects a series of parallel elements for
+    indexing together, stopping before another heading
+    """
+
+    # Handle no sibling
+    if not sibling:
+        return ''
+
+    # Handle another heading
+    if sibling.name == original_part.name:
+        return ''
+    if isinstance(sibling, Tag) and sibling.find(original_part.name):
+        return ''
+
+    # Recursively get sibling text
+    if isinstance(sibling, NavigableString):
+        sibling_text = sibling.strip()
+    else:
+        sibling_text = sibling.get_text().strip()
+    next_sibling_text = gobble_sibling_text(
+        sibling.next_sibling,
+        original_part
+    )
+
+    # Decompose
+    if isinstance(sibling, NavigableString):
+        sibling.replace_with('')
+    else:
+        sibling.decompose()
+
+    gobbled_text = sibling_text + ' ' + next_sibling_text
+    return gobbled_text.strip()
+
+
+def get_text_for_parent(parent, original_part):
+    if not parent:
+        return ''
+
+    text = gobble_sibling_text(parent.next_sibling, original_part)
+    if len(text.strip()) > len(original_part.get_text().strip()):
+        return text
+    else:
+        return get_text_for_parent(parent.parent, original_part)
+
+
+def get_text_for_header(part):
+    """
+    Gets the sibling text or parent text
+    when given an h2, h3, or h4
+    """
+
+    # Try gobbling first
+    text = gobble_sibling_text(part.next_sibling, part)
+    if len(text.strip()) > 0:
+        return text
+
+    # Otherwise go to parent
+    return get_text_for_parent(part.parent, part)
+
+
+def add_part_as_doc(part, part_url, part_name, part_text):
+    if not part_name or not part_text:
+        # HTML tree with id has no heading
+        # or no text content
+        return
+    add_search_index_document(part_url, part_name, part_text)
+    part.decompose()
+
+
+def add_searchable_page_parts(url, body):
+
+    headings = ['h4', 'h3', 'h2']
+
+    def get_part_url(part):
+        part_id = part.get('id', '')
+        if part_id:
+            return urljoin(url, '#' + part_id)
+        else:
+            return remove_fragment(url)
```

Comment:
why is this function nested? it doesn't seem to make use of parent state as say a closure or decorator

Reply:
For context, but I'll take it out